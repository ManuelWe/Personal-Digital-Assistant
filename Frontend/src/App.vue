<template>
  <section class="hero is-info is-fullheight">
    <div class="hero-head">
      <Navbar>
        <SpeechRecognition
          ref="speechRecognitionComponent"
          :user-input="userInput"
          @update:user-input="userInput = $event"
        />
      </Navbar>
    </div>

    <router-view
      ref="routerView"
      :user-input="userInput"
      @update:user-input="userInput = $event"
    />
  </section>
</template>

<script>
import SpeechService from '@/services/SpeechSynthesis';
import Navbar from './components/Navbar.vue';
import SpeechRecognition from './components/SpeechRecognition.vue';
import SpeechRecognitionLogic from './mixins/SpeechRecognitionLogic';
import LocationService from './services/Location';

export default {
  components: {
    Navbar,
    SpeechRecognition,
  },
  mixins: [SpeechRecognitionLogic],
  data() {
    return {
      userInput: '',
    };
  },
  watch: {
    // eslint-disable-next-line func-names
    '$route.name': function () {
      this.userInput = '';
    },
  },
  created() {
    this.$buefy.dialog.confirm({
      message: 'This Page talks to you!',
      type: 'is-success',
      onConfirm: () => {
        if (localStorage.getItem('soundEnabled') === 'true') SpeechService.speak('Hello my name is Gunter!');
      },
    });
  },
  beforeUpdate() {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(this.showPosition, this.error);
    } else {
      this.$buefy.toast.open({
        message: 'Geolocation is not supported by this browser.',
        duration: 3000,
        type: 'is-danger',
      });
    }
  },
  methods: {
    showPosition(position) {
      LocationService.sendPosition({
        lat: position.coords.latitude,
        lon: position.coords.longitude,
      }).catch((error) => this.$buefy.toast.open({
        message: `Error ${error.response.data.status}: ${error.response.data.error}`,
        duration: 3000,
        type: 'is-danger',
      }));
    },
    error(err) {
      this.$buefy.toast.open({
        message: `ERROR(${err.code}): ${err.message}`,
        duration: 3000,
        type: 'is-danger',
      });
    },
  },
};
</script>

<style scoped>
.hero.is-info {
  background: linear-gradient(
      rgba(0, 0, 0, 0.5),
      rgba(0, 0, 0, 0.5)
    ), url('./assets/wallpaper.jpg') no-repeat center center fixed;
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover;
}
</style>
