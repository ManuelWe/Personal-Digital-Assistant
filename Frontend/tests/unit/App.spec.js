import { createLocalVue, mount } from '@vue/test-utils';
import Buefy from 'buefy';
import VueRouter from 'vue-router';
import mockAxios from 'axios';
import persistentState from 'vue-persistent-state';
import SpeechService from '@/services/SpeechSynthesis';
import App from '@/App.vue';

const localVue = createLocalVue();
localVue.use(Buefy);
localVue.use(VueRouter);
localVue.use(persistentState, {
  soundEnabled: true,
  microphoneEnabled: true,
  notificationsEnabled: false,
  notificationEndpoint: '',
});

const routes = [
  {
    path: '/',
    name: 'landingPage',
    component: App,
  },
  {
    path: '/dialog',
    name: 'dialog',
    component: App,
  },
];
const router = new VueRouter({
  routes,
});

jest.mock('@/services/SpeechSynthesis');
jest.mock('axios', () => ({
  put: jest.fn(() => Promise.reject()),
  create: () => mockAxios,
  defaults: {
    adapter: {},
  },
}));

function factory() {
  return mount(App, {
    localVue,
    router,
    stubs: ['Navbar'],
  });
}


describe('App.vue', () => {
  describe('Elements are rendered correctly', () => {
    it('Component renders', () => {
      const wrapper = factory();
      expect(wrapper).toMatchSnapshot();
    });

    it('Navbar component renders', () => {
      const wrapper = factory();
      expect(wrapper.find('navbar-stub').isVisible()).toBe(true);
    });
  });

  it('position is put successfully', () => {
    const wrapper = factory();
    wrapper.vm.showPosition({ coords: { latitude: '21.343234', longitude: '12.345344' } });
    wrapper.vm.$nextTick(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('error message is shown', () => {
    const wrapper = factory();
    wrapper.vm.error('ERROR');
    wrapper.vm.$nextTick(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('test SpeechRecognitionLogic.js mixin', () => {
    it('route change, when userInput equals keyword', () => {
      const keywords = ['trainer', 'commute', 'restaurant', 'travel'];
      const wrapper = factory();
      for (let i = 0; i < keywords.length; i += 1) {
        wrapper.vm.checkForUseCase(keywords[i]);
        expect(router.currentRoute.name).toBe('dialog');
      }
    });

    it('time is read, when user inputs time keyword', () => {
      const wrapper = factory();
      wrapper.vm.checkForUseCase('time');
      expect(SpeechService.speak).toHaveBeenCalled();
    });

    it('userInput watcher calls function', () => {
      const wrapper = factory();
      const spy = jest.spyOn(wrapper.vm, 'checkForUseCase');
      wrapper.vm.userInput = 'Test';
      wrapper.vm.$nextTick(() => {
        expect(spy).toHaveBeenCalled();
      });
    });
  });
});
