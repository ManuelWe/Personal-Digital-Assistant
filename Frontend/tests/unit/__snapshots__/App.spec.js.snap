// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue Elements are rendered correctly Component renders 1`] = `
<section class="hero is-info is-fullheight">
  <div class="hero-head">
    <navbar-stub>
      <div id="speechRecognition"><span data-label="Mute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"></span> <span data-label="Unmute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button id="unmuteButton" class="button is-info"><span class="icon"><i class="fas fa-microphone-slash fa-2x"></i></span></button></span></div>
    </navbar-stub>
  </div>
  <section class="hero is-info is-fullheight" user-input="">
    <div class="hero-head">
      <navbar-stub>
        <div id="speechRecognition"><span data-label="Mute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"></span> <span data-label="Unmute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button id="unmuteButton" class="button is-info"><span class="icon"><i class="fas fa-microphone-slash fa-2x"></i></span></button></span></div>
      </navbar-stub>
    </div>
    <!---->
  </section>
</section>
`;

exports[`App.vue error message is shown 1`] = `
<section class="hero is-info is-fullheight">
  <div class="hero-head">
    <navbar-stub>
      <div id="speechRecognition"><span data-label="Mute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"></span> <span data-label="Unmute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button id="unmuteButton" class="button is-info"><span class="icon"><i class="fas fa-microphone-slash fa-2x"></i></span></button></span></div>
    </navbar-stub>
  </div>
  <section class="hero is-info is-fullheight" user-input="">
    <div class="hero-head">
      <navbar-stub>
        <div id="speechRecognition"><span data-label="Mute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"></span> <span data-label="Unmute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button id="unmuteButton" class="button is-info"><span class="icon"><i class="fas fa-microphone-slash fa-2x"></i></span></button></span></div>
      </navbar-stub>
    </div>
    <!---->
  </section>
</section>
`;

exports[`App.vue position is put successfully 1`] = `
<section class="hero is-info is-fullheight">
  <div class="hero-head">
    <navbar-stub>
      <div id="speechRecognition"><span data-label="Mute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"></span> <span data-label="Unmute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button id="unmuteButton" class="button is-info"><span class="icon"><i class="fas fa-microphone-slash fa-2x"></i></span></button></span></div>
    </navbar-stub>
  </div>
  <section class="hero is-info is-fullheight" user-input="">
    <div class="hero-head">
      <navbar-stub>
        <div id="speechRecognition"><span data-label="Mute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"></span> <span data-label="Unmute" class="is-light is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button id="unmuteButton" class="button is-info"><span class="icon"><i class="fas fa-microphone-slash fa-2x"></i></span></button></span></div>
      </navbar-stub>
    </div>
    <!---->
  </section>
</section>
`;
